cvat-sdk/cvat_sdk/pytorch/caching.py

```diff
@@ -0,0 +1,213 @@
+
+    No network access will be performed if this policy is used.
+    """
+
+
+_ModelType = TypeVar("_ModelType", bound=OpenApiModel)
+
+
+class CacheManager:
+    def __init__(self, client: cvat_sdk.core.Client) -> None:
+        self._client = client
+        self._logger = client.logger
+
+        # Base64-encode the name to avoid FS-unsafe characters (like slashes)
+        server_dir_name = (
+            base64.urlsafe_b64encode(client.api_map.host.encode()).rstrip(b"=").decode()
+        )
+        self._server_dir = client.config.cache_dir / f"servers/{server_dir_name}"
+
+    def task_dir(self, task_id: int) -> Path:
+        return self._server_dir / f"tasks/{task_id}"
+
+    def task_json_path(self, task_id: int) -> Path:
+        return self.task_dir(task_id) / "task.json"
+
+    def chunk_dir(self, task_id: int) -> Path:
+        return self.task_dir(task_id) / "chunks"
+
+    def project_dir(self, project_id: int) -> Path:
+        return self._server_dir / f"projects/{project_id}"
+
+    def project_json_path(self, project_id: int) -> Path:
+        return self.project_dir(project_id) / "project.json"
+
+    def load_model(self, path: Path, model_type: Type[_ModelType]) -> _ModelType:
+        with open(path, "rb") as f:
+            return model_type._new_from_openapi_data(**json.load(f))
+
+    def save_model(self, path: Path, model: OpenApiModel) -> None:
+        with atomic_writer(path, "w", encoding="UTF-8") as f:
+            json.dump(to_json(model), f, indent=4)
+            print(file=f)  # add final newline
+
+    def retrieve_task(self, task_id: int) -> Task:
+        raise NotImplementedError
+
+    def ensure_task_model(
+        self,
+        task_id: int,
+        filename: str,
+        model_type: Type[_ModelType],
+        download: Callable[[], _ModelType],
+        model_description: str,
+    ) -> _ModelType:
+        raise NotImplementedError
+
+    def ensure_chunk(self, task: Task, chunk_index: int) -> None:
+        raise NotImplementedError
+
+    def retrieve_project(self, project_id: int) -> Project:
+        raise NotImplementedError
+
+
+class _CacheManagerOnline(CacheManager):
+    def retrieve_task(self, task_id: int) -> Task:
+        self._logger.info(f"Fetching task {task_id}...")
+        task = self._client.tasks.retrieve(task_id)
+
+        self._initialize_task_dir(task)
+        return task
+
+    def _initialize_task_dir(self, task: Task) -> None:
+        task_dir = self.task_dir(task.id)
+        task_json_path = self.task_json_path(task.id)
+
+        try:
+            saved_task = self.load_model(task_json_path, models.TaskRead)
+        except Exception:
+            self._logger.info("Task is not yet cached or the cache is corrupted")
+
+            # If the cache was corrupted, the directory might already be there; clear it.
+            if task_dir.exists():
+                shutil.rmtree(task_dir)
+        else:
+            if saved_task.updated_date < task.updated_date:
```

Comment:
It might be better to compare dates wrt time zone.

Reply:
Isn't that already how `datetime` comparison works?

Reply:
Ok, you're right, it's already being used.